### PR TITLE
feat(FEV-886): Drag and Snap + Animation

### PR DESCRIPTION
**Pull request by semarche-kaltura**
animation according to https://kaltura.atlassian.net/browse/FEV-895

### CODE REVIEW

File: src/components/pip-minimized/pip-minimized.tsx [hunk 1]
```diff
@@ -1,17 +1,32 @@
...
+  redux: {connect},
+  components: {PLAYER_SIZE, Icon}
+} = KalturaPlayer.ui;
 
-interface PIPMinimizedProps {
+enum InteractiveAreaMargin {
+  DEFAULT = '16px',
+  SMALL = '8px'
+}
+
```
Comment (RoyBregman) [inline src/components/pip-minimized/pip-minimized.tsx:13]: should be handled by core already. please take latest core and see

File: src/components/pip-minimized/pip-minimized.tsx [hunk 2]
```diff
@@ -1,17 +1,32 @@
...
+interface PIPMinimizedConnectProps {
+  playerSize?: string;
+}
+
+type PIPMinimizedProps = PIPMinimizedOwnProps & PIPMinimizedConnectProps;
 
+const mapStateToProps = (state: Record<string, any>) => ({
+  playerSize: state.shell.playerSize
+});
+@connect(mapStateToProps)
```
Comment (RoyBregman) [inline src/components/pip-minimized/pip-minimized.tsx:29]: might be redundant if core makes the responsivness

File: src/components/pip/pip.scss [hunk 3]
```diff
@@ -1,24 +1,23 @@
...
-    overflow: hidden;
-    box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.2), 0 0 4px 0 rgba(0, 0, 0, 0.3);
+  &:not(.dragging) {
+    transition: all $animation-duration ease-in;
   }
 
-  &.bottom {
-    bottom: 0;
+  &.playerHover {
+    margin-top: 50px;
```
Comment (RoyBregman) [inline src/components/pip/pip.scss:14]: this should be 37px ( calculated by vars of button height + button gap)

File: src/components/pip/pip.scss [hunk 4]
```diff
@@ -1,24 +1,23 @@
...
+
 .childPlayer {
   position: absolute;
 
-  .videoContainer {
-    border-radius: 8px;
-    overflow: hidden;
-    box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.2), 0 0 4px 0 rgba(0, 0, 0, 0.3);
+  &:not(.dragging) {
+    transition: all $animation-duration ease-in;
```
Comment (RoyBregman) [inline src/components/pip/pip.scss:10]: check if all does't do performance issue. Maybe better to specific

File: src/dualscreen.tsx [hunk 5]
```diff
@@ -266,7 +289,7 @@ export class DualScreen extends KalturaPlayer.core.BasePlugin {
         label: 'kaltura-dual-screen-side-by-side',
         presets: ['Playback', 'Live', 'Error', 'Ads', 'Idle'],
```
Comment (RoyBregman) [inline src/dualscreen.tsx:290]: leave only playback and live. (ads?)

File: src/components/pip/pip.scss [hunk 6]
```diff
@@ -1,24 +1,23 @@
...
   }
 
-  &.right {
-    right: 0;
-  }
-  &.left {
-    left: 0;
+  .videoContainer {
+    border-radius: 8px;
+    overflow: hidden;
```
Comment (RoyBregman) [inline src/components/pip/pip.scss:19]: use maybe general css

File: src/components/pip/pip.tsx [hunk 7]
```diff
@@ -1,14 +1,19 @@
 import {h, createRef, Component} from 'preact';
 import * as styles from './pip.scss';
-import {Position} from '../../enums/positionEnum';
+import {Position, Animaitons} from '../../enums';
```
Comment (RoyBregman) [inline src/components/pip/pip.tsx:3]:
```suggestion
import {Position, Animations} from '../../enums';
```

File: src/components/pip/pip.tsx [hunk 8]
```diff
@@ -19,36 +24,68 @@ interface PIPComponentOwnProps {
...
+    if (dragAndSnapManager && guiClientRect) {
+      dragAndSnapManager.setGuiClientRect(guiClientRect);
+    }
+  }
+
+  private _handleClick = (fn: Function) => (e: Event) => {
+    e.stopPropagation();
+    this.props.dragAndSnapManager?.destroy();
+    fn();
+  };
```
Comment (RoyBregman) [inline src/components/pip/pip.tsx:71]: move to utils

Comment (RoyBregman) [reply src/components/pip/pip.tsx:71]: and add to all buttons

File: src/utils/index.ts [hunk 9]
```diff
@@ -1,3 +1,21 @@
...
+};
+
+export const getClientY = (e: MouseEvent | TouchEvent): number => {
+  if (e instanceof MouseEvent) {
+    return e.clientY;
+  }
+  return e.changedTouches && e.changedTouches[0] && e.changedTouches[0].clientY;
+};
+
+export const makeStyleString = (position: number): string => {
```
Comment (RoyBregman) [inline src/utils/index.ts:19]:
```suggestion
export const dimensionStyleToString = (position: number): string => {
```

File: src/components/pip/pip.tsx (hunk 8 above)
Comment (RoyBregman) [reply src/components/pip/pip.tsx:71]: Like discussed - better to make a new HOC to wrap all buttons and handle clicks, tooltip, and more

File: src/dualscreen.tsx [hunk 10]
```diff
@@ -34,9 +34,10 @@ export class DualScreen extends KalturaPlayer.core.BasePlugin {
     this._player = player;
     this._addBindings();
     this._secondaryKalturaPlayer = this._createDummyChildPlayer();
-    this._secondaryKalturaPlayer.loadMedia({entryId: '1_q31s7nkk'});
+    this._secondaryKalturaPlayer.loadMedia({entryId: '0_4rvdnj48'});
```
Comment (yairans) [inline src/dualscreen.tsx:37]: make sure to remove it after the provider PR merged

File: src/dualscreen.tsx [hunk 11]
```diff
@@ -300,8 +303,8 @@ export class DualScreen extends KalturaPlayer.core.BasePlugin {
         disable: true
       },
       provider: {
-        partnerId: 811441,
-      },
+        partnerId: 811441
```
Comment (yairans) [inline src/dualscreen.tsx:306]: remove

**Review by yairans — CHANGES_REQUESTED** [review]

File: src/utils/index.ts (hunk 9 above)
Comment (semarche-kaltura) [reply src/utils/index.ts:19]: done

File: src/dualscreen.tsx (hunk 5 above)
Comment (semarche-kaltura) [reply src/dualscreen.tsx:274]: done

File: src/components/pip/pip.scss (hunk 3 above)
Comment (semarche-kaltura) [reply src/components/pip/pip.scss]: there is an issue with interactive area, let's discuss how and where we gonna fix current bug

File: src/components/pip/pip.scss (hunk 6 above)
Comment (semarche-kaltura) [reply src/components/pip/pip.scss]: current styles relevant only for PIP component

File: src/components/pip/pip.tsx (hunk 8 above)
Comment (semarche-kaltura) [reply src/components/pip/pip.tsx]: done

File: src/components/pip/pip.tsx [hunk 12]
```diff
@@ -19,36 +24,68 @@ interface PIPComponentOwnProps {
...
+  componentDidUpdate() {
+    const {dragAndSnapManager, guiClientRect} = this.props;
+    if (dragAndSnapManager && guiClientRect) {
+      dragAndSnapManager.setGuiClientRect(guiClientRect);
+    }
+  }
+
+  private _handleClick = (fn: Function) => (e: Event) => {
+    e.stopPropagation();
+    this.props.dragAndSnapManager?.destroy();
```
[the remark this answers is not in the document]
Comment (semarche-kaltura) [reply src/components/pip/pip.tsx]: not relevant anymore

File: src/components/pip/pip.tsx [hunk 13]
```diff
@@ -19,36 +24,68 @@ interface PIPComponentOwnProps {
...
 
+  componentDidUpdate() {
+    const {dragAndSnapManager, guiClientRect} = this.props;
+    if (dragAndSnapManager && guiClientRect) {
+      dragAndSnapManager.setGuiClientRect(guiClientRect);
+    }
+  }
+
+  private _handleClick = (fn: Function) => (e: Event) => {
+    e.stopPropagation();
```
[the remark this answers is not in the document]
Comment (semarche-kaltura) [reply src/components/pip/pip.tsx]: increase performance

File: src/components/pip-minimized/pip-minimized.tsx [hunk 14]
```diff
@@ -24,18 +30,20 @@ export class PipMinimized extends Component<PIPMinimizedProps> {
...
       <Fragment>
         {!hideButtons && (
-          <div onMouseUp={show} role="button" className={styles.showContainer}>
-            <div className={styles.iconContainer}>
-              <Icon id="dualscreen-pip-minimized-show" height={16} width={16} path={icons.SHOW_ICON_PATH} />
-            </div>
-            Show
-          </div>
+          <Button onClick={show} className={styles.showContainer}>
+            <Fragment>
```
Comment (RoyBregman) [inline src/components/pip-minimized/pip-minimized.tsx:34]: Fragment is not needed if wrapped in the button

File: src/components/pip/pip-child.tsx [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+  private _renderHoverButton() {
+    const {onSideBySideSwitch, hide, onInversePIP, playerHover, isDragging} = this.props;
+    if (!playerHover || isDragging) {
+      return null;
+    }
+    return (
+      <div>
+        <div className={styles.innerButtons}>
+          <Button className={styles.iconContainer} onClick={onInversePIP}>
+            <Icon id="dualscreen-pip-swap" height={24} width={24} path={icons.SWAP_ICON_PATH} />
```
Comment (RoyBregman) [inline src/components/pip/pip-child.tsx:57]: Export the icon sizes in icons.ts and use them here instead of 24 / 16

File: src/components/pip/pip.scss [hunk 16]
```diff
@@ -31,9 +23,9 @@
 
     position: absolute;
     width: 66px;
-    height: 32px;
+    height: $hide-container-height;
     right: 0px;
-    top: -37px;
+    top: calc((5px + #{$hide-container-height}) * -1);
```
Comment (RoyBregman) [inline src/components/pip/pip.scss:28]: take the 5px also as a meaningful var (hideContainerGap or sonething)

File: src/components/side-by-side/side-by-side.scss [hunk 17]
```diff
@@ -14,5 +16,43 @@
...
+  animation: fading $animation-duration linear;
+}
+
+@keyframes appears {
+  from {
+    visibility: hidden;
+  }
+  to {
+    visibility: hidden;
+  }
```
Comment (RoyBregman) [inline src/components/side-by-side/side-by-side.scss:37]: from hidden to hidden? Isn't it a bug?

File: src/components/side-by-side/side-by-side.scss [hunk 18]
```diff
@@ -1,6 +1,8 @@
 @import '~styles/variables';
 @import '~styles/styles';
 
+$animation-duration: 0.3s;
```
Comment (RoyBregman) [inline src/components/side-by-side/side-by-side.scss:4]: Maybe define only one and use them in both side by side / pip

File: src/components/side-by-side/side-by-side.tsx [hunk 19]
```diff
@@ -34,19 +36,25 @@ export class SideBySide extends Component<SideBySideComponentProps> {
...
+  render({guiClientRect, animated}: SideBySideComponentProps) {
     const playerContainerStyles = {
       height: (guiClientRect?.width! / 2 / 16) * 9
     };
+    const classNames = [styles.secondaryPlayer];
+    if (animated) {
+      classNames.push(styles.animatedScale);
+    } else {
+      classNames.push(styles.animatedFade);
+    }
```
Comment (RoyBregman) [inline src/components/side-by-side/side-by-side.tsx:55]: Boolean name is not representing correctly as both if and else do animations of some kind so they are both "animated"

File: src/dualscreen.tsx (hunk 10 above)
Comment (RoyBregman) [inline src/dualscreen.tsx:37]: remove from the final PR as it is not part of the change and will just cause a conflict

File: src/dualscreen.tsx [hunk 20]
```diff
@@ -95,29 +96,35 @@ export class DualScreen extends KalturaPlayer.core.BasePlugin {
...
+              />
+            </DragAndSnapManager>
           </ResponsiveManager>
         )
       })
     );
   };
 
-  private _switchToPIPInverse = (manualChange: boolean = true) => {
+  private _switchToPIPInverse = (manualChange: boolean, animation: Animations) => {
```
Comment (RoyBregman) [inline src/dualscreen.tsx:127]: When manualChange is false you send animations.none to the parent pip. Child pip never uses the animation parameter - So, first the name is confusing. Maybe call it parentAnimation. Second, I would give a default "none" and don't pass it at all when manualChange is false. And then actually you can remove completly the manualChange param as it is used only for the pipParent animation decision

File: src/components/pip-minimized/pip-minimized.tsx (hunk 14 above)
Comment (semarche-kaltura) [reply src/components/pip-minimized/pip-minimized.tsx:34]: Button takes only single child node, that's why Fragment needed here.

File: src/components/side-by-side/side-by-side.scss (hunk 17 above)
Comment (semarche-kaltura) [reply src/components/side-by-side/side-by-side.scss:37]: it makes a trick with buttons that should be hidden during animation

File: src/dualscreen.tsx [hunk 21]
```diff
@@ -34,7 +34,7 @@ export class DualScreen extends KalturaPlayer.core.BasePlugin {
     this._player = player;
     this._addBindings();
     this._secondaryKalturaPlayer = this._createDummyChildPlayer();
-    this._secondaryKalturaPlayer.loadMedia({entryId: '0_4rvdnj48'});
+    this._secondaryKalturaPlayer.loadMedia({entryId: '1_q31s7nkk'});
```
Comment (RoyBregman) [inline src/dualscreen.tsx:37]: remove from PR

**Review by RoyBregman — APPROVED** [review]

**Review by yairans — APPROVED** [review]
